mdal/mdal_data_model.hpp

```diff
@@ -74,42 +71,94 @@ namespace MDAL
   {
     public:
       std::string getMetadata( const std::string &key );
-
       void setMetadata( const std::string &key, const std::string &val );
 
       std::string name();
       void setName( const std::string &name );
 
       Metadata metadata;
-
-      bool isScalar = true;
-      bool isOnVertices = true;
       Datasets datasets;
-      std::string uri; // file/uri from where it came
+      Mesh *parent = nullptr;
+
+      bool isScalar() const;
+      void setIsScalar( bool isScalar );
+
+      bool isOnVertices() const;
+      void setIsOnVertices( bool isOnVertices );
+
+      std::string uri() const;
+      void setUri( const std::string &uri );
+
+      Statistics statistics() const;
+      void setStatistics( const Statistics &statistics );
+
+    private:
+      bool mIsScalar = true;
+      bool mIsOnVertices = true;
+      std::string mUri; // file/uri from where it came
+      Statistics mStatistics;
   };
 
   typedef std::vector<std::shared_ptr<DatasetGroup>> DatasetGroups;
 
-  struct Mesh
+  class MeshVertexIterator
   {
-    std::string uri; // file/uri from where it came
-    std::string crs;
-
-    Vertices vertices;
-    std::map<size_t, size_t> vertexIDtoIndex; // only for 2DM and DAT files
-
-    Faces faces;
-    std::map<size_t, size_t> faceIDtoIndex; // only for 2DM and DAT files
+    public:
+      virtual ~MeshVertexIterator();
 
-    DatasetGroups datasetGroups;
+      virtual size_t next( size_t vertexCount, double *coordinates ) = 0;
+  };
 
-    void setSourceCrs( const std::string &str );
-    void setSourceCrsFromWKT( const std::string &wkt );
-    void setSourceCrsFromEPSG( int code );
+  class MeshFaceIterator
+  {
+    public:
+      virtual ~MeshFaceIterator();
 
-    void addBedElevationDataset();
+      virtual size_t next( size_t faceOffsetsBufferLen,
+                           int *faceOffsetsBuffer,
+                           size_t vertexIndicesBufferLen,
+                           int *vertexIndicesBuffer ) = 0;
   };
 
+  class Mesh
+  {
+    public:
+      Mesh( size_t verticesCount,
+            size_t facesCount,
+            size_t faceVerticesMaximumCount,
+            BBox extent,
+            const std::string &uri );
+      virtual ~Mesh();
+
+      void setSourceCrs( const std::string &str );
+      void setSourceCrsFromWKT( const std::string &wkt );
+      void setSourceCrsFromEPSG( int code );
+
+      void setVerticesCount( const size_t &verticesCount );
+      void setFacesCount( const size_t &facesCount );
+      void setFaceVerticesMaximumCount( const size_t &faceVerticesMaximumCount );
```

Comment:
better to pass simple values by value than by const ref

Reply:
that is how it ends up when you trust automatic code creation tools ...